### DISCUSSION TITLE
Discriminate union based on field existing/not existing

### DISCUSSION THREAD

**Discussion by Ran4** [pydantic/pydantic#4029] [Question]
I have two BaseModels, where the only difference is that one of them has a field while the other does not.

How do I correctly specify this discriminator?

Example code:

```python
from typing import Dict, Union
from uuid import uuid4, UUID

from pydantic import BaseModel


class PersonV1(BaseModel):
    pass


class PersonV2(BaseModel):
    name: str


class PersonGroup(BaseModel):
    people: Dict[UUID, Union[PersonV1, PersonV2]]


# This will print
# people={
#   UUID('a1385046-1959-437d-8e44-159a53349ece'): PersonV1()
# }
# even though we created a PersonV2
print(
    PersonGroup(
        people={
            uuid4(): PersonV2(name="John"),
        }
    )
)
```

I could modify PersonV1 by adding for example `version: Literal["v1"] = "v1"` (and do the same for PersonV2) which would work (or, explicitly, by replacing `Union[PersonV1, PersonV2]` with `Annotated[Union[PersonV1, PersonV2], Field(discriminator="version")]`).

but in my case, my (real) PersonV1 is part of a public api that I can't easily change.

Intuitively I'd like to do an annotation with `Field(discriminator=lambda person: getattr(person, "name", None)` or something like that, but `discriminator` only takes a `str`, not a callable.

**Comment by gryevns** [comment]
Not sure if this example is simplified and missing details but you could swap the order of classes?
```python
class PersonGroup(BaseModel):
    people: Dict[UUID, Union[PersonV2, PersonV1]]
```
> As such, it is recommended that, when defining Union annotations, the most specific type is included first and followed by less specific types.

https://pydantic-docs.helpmanual.io/usage/types/#unions
